### PR TITLE
NLL transition

### PR DESCRIPTION
**Pull request by nikomatsakis**
Blog post to make sure people are aware that we are moving NLL-related warnings to hard errors, and they may have to upgrade their dependencies.

Opening paragraph:

> For most of 2018, we've been issue warnings about various bugs in the
borrow checker that we plan to fix -- about two months ago, in the
current Rust nightly, those warnings became **hard errors**. In about
two weeks, when the nightly branches to become beta, those hard errors
will be in the beta build, and they will eventually hit stable on
December 19th, as part of Rust 1.40.0. **If you're testing with
Nightly, you should be all set -- but otherwise, you may want to go
and check to make sure your code still builds. If not, we have advice
for fixing common problems below.**

cc @Mark-Simulacrum @Centril @pnkfelix @matthewjasper
cc @rust-lang/compiler @rust-lang/core

### CODE REVIEW

File: posts/2019-10-28-nll-hard-errors.md [hunk 1]
```diff
@@ -0,0 +1,106 @@
+---
+layout: post
+title: "Completing the transition to the new borrow checker"
+author: Niko Matsakis
+---
+
+For most of 2018, we've been issue warnings about various bugs in the
```
Comment (Mark-Simulacrum) [inline posts/2019-10-28-nll-hard-errors.md]:
```suggestion
For most of 2018, we've been issuing warnings about various bugs in the
```

File: posts/2019-10-28-nll-hard-errors.md [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+available that *do* compile (for the most part, this had all already
+happened during the warning period). But if you have those older
+versions in your `Cargo.lock` file, and you are only using stable
+builds, then you may find that your code no longer builds once 1.40.0
+is released -- you will have to upgrade the dependency.
+
+The most common crates that were affected are the following:
+
+* `url` version 1.7.0 -- you can upgrade to 1.7.2, though you'd be better off upgrading to 2.1.0
+* `nalgebra` version 0.16.13 -- you can upgrade to 0.16.14, though you'd be better off upgrading to 0.19.0
```
Comment (Mark-Simulacrum) [inline posts/2019-10-28-nll-hard-errors.md]:
```suggestion
* `nalgebra` version 0.16.13 -- you can upgrade to 0.16.14, though you'd be better off upgrading to 0.19.0
* `rusttype` version 0.2.0 to 0.2.3 -- you can upgrade to 0.2.4, though you'd be better off looking at upgrading to 0.8.1
```

**Comment by Mark-Simulacrum** [comment]
Going to loosely assign myself but I think this is good to go when we're ready for it. I'd like to target this Thursday (after compiler team meeting).

File: posts/2019-10-28-nll-hard-errors.md [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+changes gradually. We usually begin with "Future Compatibility
+Warnings", for example, before moving those warnings to hard errors
+(sometimes a small bit at a time). Since the bug fixes to the borrow
+checker affected a lot of crates, we knew we needed a warning period
+before we could make them into hard errors.
+
+To implement this warning period, we kept two copies of the borrow
+checker around (this is a trick we use quite frequently, actually).
+The new checker ran first. If it found errors, we didn't report them
+directly: instead, we ran the old checker in order to see if the crate
```
Comment (Centril) [inline posts/2019-10-28-nll-hard-errors.md]:
```suggestion
directly: instead, we ran the old checker to see if the crate
```

File: posts/2019-10-28-nll-hard-errors.md [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+who was using Rust 2015 can attest that this shift was a big
+improvement.
+
+### The new borrow checker also fixed a lot of bugs
+
+What is perhaps less well understood is that the new borrow checker
+implementation *also* fixed a lot of bugs. In other words, the new
+borrow checker did not just accept more programs -- **it also rejected
+some programs that never should have been accepted in the first
+place!**
```
Comment (Centril) [inline posts/2019-10-28-nll-hard-errors.md]: Maybe mention that these were soundness bugs ("memory unsafety bugs" might work better).

File: posts/2019-10-28-nll-hard-errors.md [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+
+* `url` version 1.7.0 -- you can upgrade to 1.7.2, though you'd be better off upgrading to 2.1.0
+* `nalgebra` version 0.16.13 -- you can upgrade to 0.16.14, though you'd be better off upgrading to 0.19.0
+* `rusttype` version 0.2.0 to 0.2.3 -- you can upgrade to 0.2.4, though you'd be better upgrading to 0.8.1
+
+You can find out which crates you rely upon using the [cargo-tree] command. If you find
+that you *do* rely (say) on `url` 1.7.0, you can upgrade to 1.7.2 by executing:
+
+```bash
+cargo update -p url --precise 1.7.2
```
Comment (tshepang) [inline posts/2019-10-28-nll-hard-errors.md]: can you change `-p` to `--package` so it's more clear

File: posts/2019-10-28-nll-hard-errors.md (hunk 3 above)
Comment (nikomatsakis) [reply posts/2019-10-28-nll-hard-errors.md]: I'm going to push back here. I realize "in order" is not necessary, but I find it works in the flow of this sentence for reasons I find hard to *precisely* articulate. Somehow "ran the old checker to see if the crate used to compile" feels like it "packs a lot in", requiring more mental effort for me to parse and understand, whereas "in order to see" clarifies the overall structure of the sentence and is easier.

Comment (steveklabnik) [reply posts/2019-10-28-nll-hard-errors.md]: I think this is fine as-is, and generally prefer the author's preference when it comes to stuff like this.

Comment (Centril) [reply posts/2019-10-28-nll-hard-errors.md]: It's your blog post, but personally I find wordiness bad in technical writing. See also:

- https://www.oxford-royale.com/articles/clumsy-phrases-avoid.html
- https://blog.apastyle.org/apastyle/2015/09/principles-of-writing-how-to-avoid-wordiness.html
- https://proofreadingpal.com/proofreading-pulse/essays/how-to-avoid-wordiness/
- https://www.bcit.ca/files/ltc/pdf/avoid_in_order_to.pdf

**Pull request merged**